docs/reference/mapping/params/subobjects.asciidoc

```diff
@@ -105,3 +110,70 @@ PUT my-index-000001/_doc/metric_1
 <2> The document does not support objects
 
 The `subobjects` setting for existing fields and the top-level mapping definition cannot be updated.
+
+==== Auto-flattening object mappings
+
+It is generally recommended to define the properties of an object that is configured with `subobjects: false` with dotted field names
+(as shown in the first example).
+However, it is also possible to define these properties as sub-objects in the mappings.
+In that case, the mapping will be automatically flattened before it is stored.
+This makes it easier to re-use existing mappings without having to re-write them.
+
+Note that auto-flattening will not work when certain <<mapping-params, mapping parameters>> are set
+on object mappings that are defined under an object configured with `subobjects: false`:
+
+* The <<enabled, `enabled`>> mapping parameter must not be `false`.
+* The <<dynamic, `dynamic`>> mapping parameter must not contradict the implicit or explicit value of the parent.
+* The <<subobjects, `subobjects`>> mapping parameter mot not be set to `true` explicitly.
```

Comment:
Think of a component template (A) that was written with "traditional" object mappings in mind (subobjects is not set, the default value true is used). This PR makes it so that this component template can be used in combination with another component template (B) that sets subobjects to false. However, one limitation is that in component template A, the object mappers must not be defined with `subobjects: true`. We could also choose to just discard the subobjects mapping parameter but I chose a more conservative approach that would reject such mappings. However, the consequence is that adding subobjects: false to existing mappings may fail.

Reply:
ok. I see how this makes sense applied to component templates. I got confused because in the docs there is no context around component templates, and subobjects can't normally be changed back to true once set to false, so I was wondering: that is already not valid mappings, why do we need to state that explicitly. I am still uncertain of whether we should specify it or not.